### PR TITLE
Lab05 Komzolova Daria

### PR DESCRIPTION
**Pull request by DariaKomzolova**
## Overview

This lab focuses on implementing secure authentication in a cross-platform application using Go for backend (REST API) and Flutter for frontend (UI, session logic, validation). It covers concepts such as JWT token generation/validation, clean architecture, user authentication state, form validation, and test-driven development.

## PR with points

https://github.com/DariaKomzolova/sum25-go-flutter-course/pull/8/

### Go Backend Component

- [✔️] JWT service with HS256 signing and 24-hour expiration  
- [✔️] User repository with mock and in-memory support  
- [✔️] RESTful handlers for login, token validation, health check  
- [✔️] Custom error handling for invalid credentials, session expiry, etc.  
- [✔️] Unit tests for all services and handlers  

### Flutter Frontend Component

- [✔️] `AuthService` class with login, logout, session validation, and JWT logic  
- [✔️] `FormValidator` with static methods for email/password/sanitize  
- [✔️] `User` domain entity with validation methods  
- [✔️] State management via `AuthState` and `copyWith`  
- [✔️] Mockable interfaces: `JWTServiceInterface`, `UserRepositoryInterface`  
- [✔️] Full unit test suite for all edge cases (invalid input, expired token, user not found, etc.)  

### Backend Testing

```bash
cd labs/lab05/backend/
go mod tidy
go test ./...
```

✅ Expected: All backend tests pass

### Frontend Testing

```bash
cd labs/lab05/frontend/
flutter pub get
flutter test
```

✅ Expected: All frontend unit tests pass (including AuthService, FormValidator, storage mocks)

## Checklist

- [✔️] All tests pass (`go test ./...` and `flutter test`)  
- [✔️] All validation methods conform to specs (lengths, formats, rules)  
- [✔️] AuthService covers login/logout/refresh/getUserInfo  
- [✔️] Mock services and data used for full testability  
- [✔️] Error messages consistent and cleanly surfaced  
- [✔️] `flutter_secure_storage` prepared for future secure token storage  
- [✔️] Code follows clean architecture (domain, core, services separated)  
- [✔️] README completed and includes test & component overview

### CODE REVIEW

File: labs/lab05/backend/jwtservice/jwt.go [hunk 1]
```diff
@@ -2,47 +2,75 @@ package jwtservice
...
+	return token.SignedString([]byte(j.secretKey))
 }
 
-// TODO: Implement ValidateToken method
 // ValidateToken parses and validates a JWT token
-// Requirements:
-// - Check token signature with secret key
-// - Verify token is not expired
-// - Return parsed claims on success
 func (j *JWTService) ValidateToken(tokenString string) (*Claims, error) {
```
Comment (BulatGazizov-dev) [inline labs/lab05/backend/jwtservice/jwt.go:53]: Do you check if the token has expired?

File: labs/lab05/backend/userdomain/user.go [hunk 2]
```diff
@@ -17,52 +17,87 @@ type User struct {
...
 // ValidatePassword checks if password meets security requirements
 func ValidatePassword(password string) error {
-	// TODO: Implement password validation
-	// Password should be at least 8 characters
-	// Should contain at least one uppercase, lowercase, and number
-	return errors.New("not implemented")
+	if len(password) < 8 {
+		return errors.New("password must be at least 8 characters")
+	}
+	hasUpper := regexp.MustCompile(`[A-Z]`).MatchString(password)
```
Comment (BulatGazizov-dev) [inline labs/lab05/backend/userdomain/user.go:87]: Same here

**Comment by BulatGazizov-dev** [comment]
You did a good job, I especially liked that you used regular expressions for validation. It would be nice to compile them only once. And try to send only one lab pull request, it will make the verification easier
